File: Sources/NotchAgent/Features/NotchOverlay/Components/SegmentedMeter.swift
import SwiftUI

/// Chunky block meter, the stick's signature gauge. Filled blocks take the
/// state color; empty blocks stay as faint sockets.
struct SegmentedMeter: View {
    let percent: Double
    var segments: Int = 14
    var tint: Color = Theme.coral
    var height: CGFloat = 9
    var spacing: CGFloat = 2

    var body: some View {
        HStack(spacing: spacing) {
            ForEach(0..<segments, id: \.self) { index in
                RoundedRectangle(cornerRadius: 1.5, style: .continuous)
                    .fill(index < filledCount ? tint : Theme.socket)
            }
        }
        .frame(height: height)
        .animation(.easeOut(duration: 0.4), value: filledCount)
    }

    private var filledCount: Int {
        let clamped = min(max(percent, 0), 100)
        // Any usage at all lights the first block.
        let count = Int((clamped / 100 * Double(segments)).rounded())
        return clamped > 0 ? max(count, 1) : 0
    }
}

File: windows/NotchAgent.Windows/UI/SegmentedMeter.cs
using Avalonia;
using Avalonia.Controls;
using Avalonia.Media;

namespace NotchAgent.Windows.UI;

/// Chunky block meter, the stick's signature gauge — a direct port of
/// SegmentedMeter.swift. Filled blocks take the state color; empty blocks
/// stay as faint sockets.
public sealed class SegmentedMeter : Control
{
    public static readonly StyledProperty<double> PercentProperty =
        AvaloniaProperty.Register<SegmentedMeter, double>(nameof(Percent));
    public static readonly StyledProperty<IBrush> TintProperty =
        AvaloniaProperty.Register<SegmentedMeter, IBrush>(nameof(Tint), AppTheme.Brush(AppTheme.Coral));
    public static readonly StyledProperty<int> SegmentsProperty =
        AvaloniaProperty.Register<SegmentedMeter, int>(nameof(Segments), 12);

    public double Percent { get => GetValue(PercentProperty); set => SetValue(PercentProperty, value); }
    public IBrush Tint { get => GetValue(TintProperty); set => SetValue(TintProperty, value); }
    public int Segments { get => GetValue(SegmentsProperty); set => SetValue(SegmentsProperty, value); }

    static SegmentedMeter()
    {
        AffectsRender<SegmentedMeter>(PercentProperty, TintProperty, SegmentsProperty);
    }

    public override void Render(DrawingContext context)
    {
        var segments = Math.Max(1, Segments);
        var clamped = Math.Min(Math.Max(Percent, 0), 100);
        var filled = clamped > 0 ? Math.Max(1, (int)Math.Round(clamped / 100 * segments)) : 0;

        const double spacing = 2;
        var totalSpacing = spacing * (segments - 1);
        var segmentWidth = (Bounds.Width - totalSpacing) / segments;
        var socket = AppTheme.Brush(AppTheme.Socket);

        for (int i = 0; i < segments; i++)
        {
            var x = i * (segmentWidth + spacing);
            var rect = new Rect(x, 0, segmentWidth, Bounds.Height);
            var brush = i < filled ? Tint : socket;
            context.DrawRectangle(brush, null, rect, 1.5, 1.5);
        }
    }
}
